### PR TITLE
make the log facility compatible with 64-bit builds

### PR DESCRIPTION
**Pull request by npitre**
To make things clearer, log arguments now have their own type to separate
them from the sea of u32_t, and to allow for a different size on 64-bit
builds. And log records must account for 64-bit pointers, etc.

### CODE REVIEW

**Review by tyrantlizard77 — APPROVED** [review]

**Comment by npitre** [comment]
This requires PR #16586 to be applied for the Kconfig issue to be resolved.

File: include/logging/log_msg.h [hunk 1]
```diff
@@ -22,22 +22,33 @@ extern "C" {
...
+
 /** @brief Maximum number of arguments in the standard log entry.
  *
  * It is limited by 4 bit nargs field in the log message.
  */
 #define LOG_MAX_NARGS 15
 
 /** @brief Number of arguments in the log entry which fits in one chunk.*/
+#ifdef CONFIG_64BIT
+#define LOG_MSG_NARGS_SINGLE_CHUNK 4
```
Comment (nordic-krch) [inline include/logging/log_msg.h]: what was the reason for that change? to have single chunk 64 bytes long?

File: tests/subsys/logging/log_msg/src/log_msg_test.c [hunk 2]
```diff
@@ -20,11 +20,16 @@ extern struct k_mem_slab log_msg_pool;
 static const char my_string[] = "test_string";
 void test_log_std_msg(void)
```
Comment (nordic-krch) [inline tests/subsys/logging/log_msg/src/log_msg_test.c]: maybe you could rework this test a bit and create a loop that tests up to six arguments and determines expected number of chunks allocated like `(arg > LOG_MSG_NARGS_SINGLE_CHUNK) ? 2 : 1`. That would look better and would not need ifdef.

File: tests/subsys/logging/log_msg/src/log_msg_test.c [hunk 3]
```diff
@@ -20,11 +20,16 @@ extern struct k_mem_slab log_msg_pool;
...
 {
+#ifdef CONFIG_64BIT
```
Comment (nordic-krch) [inline tests/subsys/logging/log_msg/src/log_msg_test.c]: what about something like:
```
int exp_nargs_in_single_chunk = IS_ENABLED(CONFIG_64BIT) ? 4 : 3;
zassert_true(LOG_MSG_NARGS_SINGLE_CHUNK == exp_nargs_in_single_chunk , "test assumes following setting");
```

**Review by nordic-krch — CHANGES_REQUESTED** [review]
Look ok (minor comments). Does string formatting works with that or does it need additional changes?

**Comment by npitre** [comment]
On Sun, 9 Jun 2019, Krzysztof Chruscinski wrote:

> Look ok (minor comments). Does string formatting works with that or does it need additional changes?

String formatting requires its own set of changes.

>  /** @brief Number of arguments in the log entry which fits in one chunk.*/
> +#ifdef CONFIG_64BIT
> +#define LOG_MSG_NARGS_SINGLE_CHUNK 4
> 
> what was the reason for that change? to have single chunk 64 bytes long?

This determines the size of union log_msg_head_data which has to match 
the size of struct log_msg_ext_head_data not to waste memory as they 
occupy the same payload space in struct log_msg. There is a build-time 
assertion enforcing that. And the size of struct log_msg_ext_head_data 
gets bigger due to the pointer and the argument type it contains which 
are larger on 64-bit targets.

> > @@ -20,11 +20,16 @@ extern struct k_mem_slab log_msg_pool;
>  static const char my_string[] = "test_string";
>  void test_log_std_msg(void)
> 
> maybe you could rework this test a bit and create a loop that tests up 
> to six arguments and determines expected number of chunks allocated 
> like `(arg > LOG_MSG_NARGS_SINGLE_CHUNK) ? 2 : 1`. That would look 
> better and would not need ifdef.

Indeed, I like your suggestion.

I'm adding an extra patch to this PR implementing those changes 
separately, in case they have unexpected side effects.

**Review by nordic-krch — APPROVED** [review]

**Comment by aescolar** [comment]
@npitre: Note that this PR is blocked due to checkpatch errors.

**Comment by npitre** [comment]
On Thu, 20 Jun 2019, Alberto Escolar wrote:

> Note that this PR is blocked due to checkpatch errors.

I'd argue that checkpatch is wrong in this particular context.
Can it be bypassed?

**Comment by nashif** [comment]
> I'd argue that checkpatch is wrong in this particular context. Can it be bypassed?

they are not wrong, we following the linux coding style and you need to have the statement and the break on a new line in this case.

**Comment by npitre** [comment]
On Thu, 20 Jun 2019, Anas Nashif wrote:

> they are not wrong, we following the linux coding style and you need to have the statement and the break on a new line in this case.

The Linux community uses a coding style and checkpatch as guidelines, 
not as a hard requirement for patch merging. Human common sense always 
prevails over checkpatch when readability of the code is improved in 
some circumstances. Linus Torvalds himself made public statements on a 
few occasions against unconditional checkpatch enforcement as it 
sometimes leads to worse code.

**Pull request merged**